builtin/logical/database/path_creds_create.go

```diff
@@ -107,21 +101,75 @@ func (b *databaseBackend) pathCredsCreateRead() framework.OperationFunc {
 			RollbackStatements: v5.Statements{
 				Commands: role.Statements.Rollback,
 			},
-			Password:   password,
 			Expiration: expiration,
 		}
 
-		// Overwriting the password in the event this is a legacy database plugin and the provided password is ignored
+		respData := make(map[string]interface{})
+
+		// Generate the credential based on the role's credential type
+		switch role.CredentialType {
+		case v5.CredentialTypePassword:
+			generator, err := newPasswordGenerator(role.CredentialConfig)
+			if err != nil {
+				return nil, fmt.Errorf("failed to construct credential generator: %s", err)
+			}
+
+			// Fall back to database config-level password policy if not set on role
+			if generator.PasswordPolicy == "" {
+				generator.PasswordPolicy = dbConfig.PasswordPolicy
+			}
+
+			// Generate the password
+			password, err := generator.generate(ctx, b, dbi.database)
+			if err != nil {
+				b.CloseIfShutdown(dbi, err)
+				return nil, fmt.Errorf("failed to generate password: %s", err)
+			}
+
+			// Set input credential
+			newUserReq.CredentialType = v5.CredentialTypePassword
+			newUserReq.Password = password
+
+			// Set output credential
+			respData["password"] = password
+
+		case v5.CredentialTypeRSAPrivateKey:
+			generator, err := newRSAKeyGenerator(role.CredentialConfig)
+			if err != nil {
+				return nil, fmt.Errorf("failed to construct credential generator: %s", err)
+			}
+
+			// Generate the RSA key pair
+			public, private, err := generator.generate(b.GetRandomReader())
+			if err != nil {
+				return nil, fmt.Errorf("failed to generate RSA key pair: %s", err)
+			}
+
+			// Set input credential
+			newUserReq.CredentialType = v5.CredentialTypeRSAPrivateKey
+			newUserReq.PublicKey = public
+
+			// Set output credential
+			respData["rsa_private_key"] = string(private)
+		}
+
+		// Overwriting the password in the event this is a legacy database
+		// plugin and the provided password is ignored
 		newUserResp, password, err := dbi.database.NewUser(ctx, newUserReq)
 		if err != nil {
 			b.CloseIfShutdown(dbi, err)
 			return nil, err
 		}
-
-		respData := map[string]interface{}{
-			"username": newUserResp.Username,
-			"password": password,
+		respData["username"] = newUserResp.Username
+
+		// Database plugins using the v4 interface generate and return the password.
+		// If the returned password is not equal to the credential, then we need to
+		// set the credential to the password returned from the v4 database plugin.
+		if role.CredentialType == v5.CredentialTypePassword &&
```

Comment:
I'm a bit confused about this this conditional. If `role.CredentialType == v5.CredentialTypePassword`, then it looks like we always set `respData["password"]` to  `password` in [L134](https://github.com/hashicorp/vault/pull/15376/files#diff-a83220b81620363186d8af7542730e768ccab09cf9a55c30cf24a5819110177bR134), so can this case ever be true?

Reply:
Ya, this caused me some confusion as well. I'm trying to preserve some behavior that existed prior to these changes. This [comment](https://github.com/hashicorp/vault/blob/main/builtin/logical/database/path_creds_create.go#L114) in the main branch sort of mentions what the issue is.

If we assume we're working with a v4 db plugin, it turns out our existing code for reaching out to the plugin to generate the password at [path_creds_create.go#L93](https://github.com/hashicorp/vault/blob/main/builtin/logical/database/path_creds_create.go#L93) is not needed. This is because the password is always generated on the db plugin side and returned in the v4 CreateUser response. Here is an example of a v4 implementation that shows this: [postgresql.go#L181](https://github.com/hashicorp/vault/blob/44fdbc7dc92ac9e9d801556c40ff9974f8dbfcd8/plugins/database/postgresql/postgresql.go#L181).

This means for v4 plugins, we're currently asking it to generate a password, and then not using it. Happy to consider how we can clean this up.

Reply:
I see. We used to always overwrite the password through since `databaseVersionWrapper.NewUser` either returns the provided password, `req.Password`, or the password returned by the plugin. I think that just the `if role.CredentialType == v5.CredentialTypePassword {...}` might be sufficient here, and we could skip setting that on [L134](https://github.com/hashicorp/vault/pull/15376/files#diff-a83220b81620363186d8af7542730e768ccab09cf9a55c30cf24a5819110177bR134)

Reply:
Ah, that's right. I didn't realize the password came back from `NewUser` for both v4 and v5. Just simplified this as you suggested in fe3e50a. Thanks!